Adds the initial support for AOTInductor model and interface

chenyang78 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103922

This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225

Commit:
Adds the initial support for AOTInductor model and interface

This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

[ghstack-poisoned]

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

chenyang78 commented:
/easycla

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

jansel commented on torch/_inductor/aot_cpp_include/aot_inductor_interface.cpp:
For this (and the other cpp files) to exist when people `pip install torch` I believe we need to update setup.py:
https://github.com/pytorch/pytorch/blob/0330f67b222bb58dcc3b47941338086e7538e704/setup.py#L1252

chenyang78 replied:
Yes, good point. Fixed.

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

chenyang78 commented on test/cpp/aot_inductor/test.py:
```diff
@@ -17,6 +17,9 @@ def forward(self, x, y):
 
+# FIXME: re-enable dynamic shape after we add dynamic shape support to the
+# AOTInductor runtime
+torch._dynamo.config.dynamic_shapes = False
```
Fixed. Thanks.

desertfire commented on test/cpp/aot_inductor/CMakeLists.txt:
```diff
@@ -30,6 +30,10 @@ target_link_libraries(test_aot_inductor PRIVATE
   ${CMAKE_CURRENT_BINARY_DIR}/libaot_inductor_output.so
 )
 
+target_include_directories(test_aot_inductor PRIVATE
```
With your change to setup.py, I suspect we don't need this now?

chenyang78 replied:
We still need this. ```torch/_inductor``` is not in the header-file search paths.

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -584,6 +587,11 @@ def cpp_flags():
 
 def optimization_flags():
     base_flags = "-O3 -ffast-math -fno-finite-math-only"
+    if config.is_fbcode():
```
You need to coordinate with Brian on this.

chenyang78 replied:
Yes, I will check with Brian. Thanks.

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -614,22 +629,32 @@ def get_include_and_linking_paths(
         # and we need a way to link to what PyTorch links.
         ipaths = cpp_extension.include_paths(cuda) + [sysconfig.get_path("include")]
+        if aot_mode:
+            ipaths += [os.path.join(_TORCH_PATH, "_inductor", "aot_cpp_include")]
```
This probably also needs to be updated after the setup.py change.

chenyang78 replied:
Yes, changed.

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -614,22 +629,32 @@ def get_include_and_linking_paths(
         # This approach allows us to only pay for what we use.
         ipaths = cpp_extension.include_paths(cuda) + [sysconfig.get_path("include")]
+        if aot_mode:
+            ipaths += [os.path.join(_INDUCTOR_PATH, "aot_cpp_include")]
```
This else branch is not tested, that's why _INDUCTOR_PATH is not defined but not caught. Honestly we don't care about this else branch, so probably ok to drop the change here.

chenyang78 replied:
Done. Thanks.

desertfire commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1086,19 +1163,31 @@ def codegen_device(self, device):
+        if output_idx is not None and V.graph.aot_mode:
+            # In aot_mode, output buffers are managed by the AOT runtime.
+            return (
+                f"at::Tensor {buffer.get_name()} = outputs[{output_idx}]{self.ending}"
```
How do we guarantee the passed-in output buffers are valid? So basically what is our protocol with the runtime?

chenyang78 replied:
The runtime will first invoke ```AOTInductorModelContainerGetMaxInputShape``` to query the output tensor shapes before allocating the right amount of memory.

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

chenyang78 commented on setup.py:
```diff
@@ -1194,6 +1194,8 @@ def main():
         'include/sleef.h',
         "_inductor/codegen/*.h",
+        "_inductor/aot_cpp_include/*.cpp",
+        "_inductor/aot_cpp_include/*.h",
```
Done. Thanks.

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

Commit:
Update on "Adds the initial support for AOTInductor model and interface"


This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

chenyang78 commented:
@pytorchbot merge

jansel commented:
The merge button doesn't work for ghstack.
